### DISCUSSION TITLE
전역 변수를 억제하는 방법에는 어떤 것이 있나요?

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#60] [전역 변수의 문제점]
- 전역 변수를 억제하는 방법에는 어떤 것이 있나요?
  - ES6 모듈 시스템이 전역 변수 문제를 어떻게 해결하나요?
  - 전역 변수의 사용을 줄이기 위한 디자인 패턴에는 어떤 것들이 있나요?

**Comment by publdaze** [comment]
> - 전역 변수를 억제하는 방법에는 어떤 것이 있나요?
>   - ES6 모듈 시스템이 전역 변수 문제를 어떻게 해결하나요?
>   - 전역 변수의 사용을 줄이기 위한 디자인 패턴에는 어떤 것들이 있나요?

- 전역 변수를 억제하는 방법에는 즉시실행함수, 모듈 패턴, 네임 스페이스 객체, ES6 모듈을 사용하는 방법이 있음
- 즉시실행함수를 사용해서 모든 코드를 감싸면 모든 코드가 즉시 실행 함수의 지역 스코프에 포함되므로 전역 변수를 억제할 수 있다.
- 모듈 패턴은 클래스를 모방한 디자인 패턴, 즉시 실행 함수로 감싸 하나의 모듈을 만드는 것인데, 클로저를 기반으로 동작하여 전역 변수 억제와 함께 캡슐화까지 구현 가능하다.
- 네임 스페이스 객체는 전역 변수처럼 사용하고 싶은 변수를 네임 스페이스 객체의 프로퍼티로 추가하는 방법, 네임스페이스 분리해서 식별자 충돌 방지
- ES6 모듈은 각 파일이 독자적인 스코프를 갖도록 제공.

**Comment by pipisebastian** [comment]
> - 전역 변수를 억제하는 방법에는 어떤 것이 있나요?
>   - ES6 모듈 시스템이 전역 변수 문제를 어떻게 해결하나요?
>   - 전역 변수의 사용을 줄이기 위한 디자인 패턴에는 어떤 것들이 있나요?


- 즉시실행함수로 감싸서, 즉시실행함수의 지역변수로 만들어주는 방법
- 디자인 패턴 - 모듈 패턴 -> 즉시실행함수 사용. 클로저를 통해 내부에서만 접근 가능.
- 네임스페이스 객체를 만들어서, 전역변수가 필요하면, 해당 네임스페이스 객체의 속성으로 추가 ⇒ 네임스페이스 자체가 전역변수긴 함. 그렇지만, 식별자 충돌은 안날거임
- es6모듈 사용하면, 전역변수를 사용할 수 없습니다. es6모듈이 (파일별로) 독자적인 스코프를 제공해줍니다.
